students/jsward/session03/mailroom.py

```diff
@@ -0,0 +1,129 @@
+                break
+    donors.append({'Name': user_input, 'Donations': []})
+    pass
+
+
+def generate_stats(donor):
+    num_donations = len(donor['Donations'])
+    total_donation = float(sum(donor['Donations']))
+    average_donation = (float(total_donation / num_donations))
+    return [donor['Name'], total_donation, num_donations, average_donation]
+
+
+def list_donors():
+    for donor in donors:
+        print(donor['Name'])
+
+
+def menu_donor():
+    while True:
+        user_input = input("l to list donors\na to add donor\n$ to add donation\nt to print thank you message\n \n").lower()
+        if user_input == 'l' or user_input == 'a' or user_input == '$' or user_input == 't':
+            return user_input
+
+
+def menu_main():
+    """
+    Prompts for user input.  Returns valid input
+    :return: string with one of the following values: q, r, t
+    """
+    while True:
+        user_input = input("q to quit\nr to generate a report\nd to manage donors (print thank you)\n \n").lower()
+        if user_input == 'q' or user_input == 'r' or user_input == 'd':
+            return user_input
+
+
+def main():
+    user_input = ''
+    while 'q' not in user_input:
+        user_input = menu_main()
+        if 'r' in user_input:
+            report()
+        elif 'd' in user_input:
+            donor_management()
+
+
+def print_thank_you(donor_name=''):
+    while True:
+        if donor_name:
+            for donor in donors:
+                if donor['Name'] == donor_name:
+                    last_donation = donor['Donations'][-1]
+                    print(
+                        "Thank you message:\n\n Dear {}, thank you for the very generous donation of ${}.  Rest assured it will be put to good use helping those in need*\n\n *After deducting 98% for management overhead, of course ;).".format(
+                            donor_name, last_donation))
+                    return
+        donor_name = str(input("Provide the name of the donor for which you'd like to print a thank you message.  "))
+
+
+def report():
+    donor_data = []
+    for donor in donors:
+        donor_data.append(generate_stats(donor))
+    sorted_donor_data = sorted(donor_data, key=itemgetter(1), reverse=True)
+    print("Donor Name{}| Total Given | Num Gifts | Average Gift".format(' ' * 20))
+    print('-' * 70)
+    for donor in sorted_donor_data:
+        total_str = '{:,.2f}'.format(donor[1])
+        average_str = '{:,.2f}'.format(donor[3])
+        print("{name}{w1}${w2}{t}{w3}{num}  ${w4}{a}".format(
+            name=donor[0],
+            w1=' ' * (31 - len(str(donor[0]))),
+            w2=' ' * (12 - len(total_str)),
+            t=total_str,
+            w3=' ' * (12 - len(str(donor[2]))),
+            num=donor[2],
+            w4=' ' * (11 - len(average_str)),
+            a=average_str
+        ))
+
+
+def donor_management():
+    user_input = menu_donor()
+    if 'l' in user_input:
+        list_donors()
+    elif 'a' in user_input:
```

Comment:
Interesting choice to split up adding a donor and adding a donation! Looks good 😄